Modules/Loadable/VolumeRendering/MRML/vtkMRMLShaderPropertyStorageNode.cxx

```diff
@@ -0,0 +1,400 @@
+/*=Auto=========================================================================
+
+Portions (c) Copyright 2005 Brigham and Women's Hospital (BWH) All Rights Reserved.
+
+See COPYRIGHT.txt
+or http://www.slicer.org/copyright/copyright.txt for details.
+
+Program:   3D Slicer
+Module:    $RCSfile: vtkMRMLShaderPropertyStorageNode.cxx,v $
+Date:      $Date: 2006/03/17 15:10:09 $
+Version:   $Revision: 1.2 $
+
+=========================================================================auto=*/
+
+#include "vtkMRMLShaderPropertyNode.h"
+#include "vtkMRMLShaderPropertyStorageNode.h"
+#include "vtkMRMLScene.h"
+
+#include <vtkObjectFactory.h>
+#include <vtkShaderProperty.h>
+#include <vtkUniforms.h>
+#include <vtkMatrix3x3.h>
+#include <vtkMatrix4x4.h>
+#include <vtkStringArray.h>
+
+#include <sstream>
+#include <json/json.h>
+
+//----------------------------------------------------------------------------
+vtkMRMLNodeNewMacro(vtkMRMLShaderPropertyStorageNode);
+
+//----------------------------------------------------------------------------
+vtkMRMLShaderPropertyStorageNode::vtkMRMLShaderPropertyStorageNode()
+{
+  this->DefaultWriteFileExtension = "sp";
+}
+
+//----------------------------------------------------------------------------
+vtkMRMLShaderPropertyStorageNode::~vtkMRMLShaderPropertyStorageNode()
+{
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLShaderPropertyStorageNode::PrintSelf(ostream& os, vtkIndent indent)
+{
+  this->Superclass::PrintSelf(os,indent);
+}
+
+//----------------------------------------------------------------------------
+bool vtkMRMLShaderPropertyStorageNode::CanReadInReferenceNode(vtkMRMLNode *refNode)
+{
+  return refNode->IsA("vtkMRMLShaderPropertyNode");
+}
+
+void DecodeValue( Json::Value & val, int & dec )
+{
+    dec = val.asInt();
+}
+
+void DecodeValue( Json::Value & val, float & dec )
+{
+    dec = val.asFloat();
+}
+
+template< typename scalarT >
+void DecodeTuple( Json::Value & val, std::vector<scalarT> & tuple )
+{
+    tuple.resize( val.size() );
+    for( unsigned i = 0; i < val.size(); ++i )
+    {
+        scalarT vi;
+        DecodeValue( val[i], vi );
+        tuple[i] = vi;
+    }
+}
+
+template< typename scalarT >
+void DecodeMatrix( Json::Value & val, std::vector<scalarT> & mat )
+{
+    int nbRows = val.size();
```

Comment:
Please fix formatting in these functions as well.

Reply:
done